python310Packages.opentelemetry: unified the source code

natsukium opened the pull request:
###### Description of changes

It was pointed out on [matrix](https://matrix.to/#/!VjfUzaKsXokUdnQcvP:nixos.org/$9DnbZKGlZacc8xm4umLfG1rsOYnd1VW8f28E6J2sEWM?via=nixos.org&via=matrix.org&via=nixos.dev) that each module under `open-telemetry/opentelemetry-python` is packaged separately and, therefore, fragile to version updates.

In this PR, I unified the `version`, `src`, and `meta` using the core package, `opentelemetry-api`, as the source.

I also added myself as a maintainer to make it easier to notice changes related to opentelemetry-api.

I hope we can package them in a better way.

Also, I will work on `opentelemetry-python-contrib` after this PR and #245028 are done.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

natsukium committed:
python310Packages.opentelemetry-api: refactor

To avoid breaking any package under the opentelemetry-python repository, we must synchronize their versions.
Since opentelemetry-api is the central component of opentelemetry-python, I used it as the source for version, src, and meta.

natsukium committed:
python310Packages.opentelemetry-api: add natsukium as maintainer

RaitoBezarius commented on pkgs/development/python-modules/opentelemetry-api/default.nix:
```diff
@@ -17,13 +17,15 @@ let
     version = "1.18.0";
     disabled = pythonOlder "3.7";
 
+    # to avoid breakage, every package in opentelemetry-python must inherit this version, src, and meta
```
It's not heriting src here, is it?

RaitoBezarius commented on pkgs/development/python-modules/opentelemetry-exporter-otlp-proto-common/default.nix:
```diff
@@ -1,26 +1,20 @@
 
-buildPythonPackage rec {
+buildPythonPackage {
+  inherit (opentelemetry-api) version src;
```
Ah I see!

RaitoBezarius commented:
Can we ping the DeShaw team too here?

natsukium commented:
> Can we ping the DeShaw team too here?

Ofcource, ofborg should already ping them. 
@limeytexan san, I'll ping you just in case.

happysalada commented:
would you mind adding grpc and aiohttp-client as well, I've just merged https://github.com/NixOS/nixpkgs/pull/245028

de11n reviewed (approved):
Thanks for doing this. I was unsure if this was the best approach when adding these packages originally, but I'm fine with doing it this way. FYI: I'm the original author of these packages and on the deshaw team, but the members are out of date (see #245907).

natsukium committed:
python310Packages.opentelemetry-instrumentation: refactor

To avoid breaking any package under the opentelemetry-python-contrib repository, we must synchronize their versions.
Since opentelemetry-instrumentation is the central component of opentelemetry-python-contrib, I used it as the source for version, src, and meta.

natsukium committed:
python310Packages.opentelemetry-instrumentation: add natsukium as maintainer

natsukium commented:
I did the same for `opentelemetry-python-contrib` and added commits.
I can drop them again if another PR is preferred.

happysalada commented:
This looks good to me, waiting one day in case anyone else wants to comment.
